File: dashboards/express/routes/journal.js
/**
 * AgenticMail Enterprise Dashboard — Journal Routes
 * GET /journal, POST /journal/:id/rollback
 */

const { Router } = require('express');
const { requireAuth } = require('../middleware/auth');
const { apiGet, apiPost } = require('../utils/api');
const { layout } = require('../views/layout');
const { buildTable } = require('../views/components/table');
const { esc, statusBadge, timeAgo } = require('../utils/helpers');

const router = Router();

router.get('/journal', requireAuth, async (req, res) => {
  const flash = req.session.flash;
  delete req.session.flash;

  const entriesResult = await apiGet('/engine/journal', req.session.token);
  const entries = entriesResult.status === 200
    ? (Array.isArray(entriesResult.body.entries) ? entriesResult.body.entries : (Array.isArray(entriesResult.body) ? entriesResult.body : []))
    : [];

  const statsResult = await apiGet('/engine/journal/stats/default', req.session.token);
  const stats = statsResult.status === 200 ? statsResult.body : {};

  const rows = entries.map(e => {
    const rollbackBtn = (e.status || '').toLowerCase() !== 'rolled_back'
      ? `<form method="post" action="/journal/${esc(e.id)}/rollback" style="display:inline" onsubmit="return confirm('Rollback this action? This cannot be undone.')">
           <button class="btn btn-sm btn-warning" type="submit">Rollback</button>
         </form>`
      : `<span class="badge badge-default">Rolled back</span>`;
    return [
      `<strong>${esc(e.action || e.type || '-')}</strong>`,
      esc(e.agent || e.agentName || '-'),
      `<code>${esc(e.resource || e.target || '-')}</code>`,
      statusBadge(e.status || 'completed'),
      `<span style="color:var(--text-muted)">${timeAgo(e.created_at || e.createdAt)}</span>`,
      rollbackBtn,
    ];
  });

  const table = buildTable(
    ['Action', 'Agent', 'Resource', 'Status', 'Time', ''],
    rows,
    '&#128216;',
    'No journal entries yet.'
  );

  const statsHtml = `
    <div class="stats-grid" style="display:grid;grid-template-columns:repeat(auto-fit,minmax(180px,1fr));gap:16px;margin-bottom:24px">
      <div class="card" style="text-align:center">
        <div style="font-size:28px;font-weight:700;color:var(--primary)">${stats.total || 0}</div>
        <div style="color:var(--text-muted);font-size:13px">Total Entries</div>
      </div>
      <div class="card" style="text-align:center">
        <div style="font-size:28px;font-weight:700;color:var(--success,#22c55e)">${stats.completed || 0}</div>
        <div style="color:var(--text-muted);font-size:13px">Completed</div>
      </div>
      <div class="card" style="text-align:center">
        <div style="font-size:28px;font-weight:700;color:var(--warning,#f59e0b)">${stats.rolled_back || stats.rolledBack || 0}</div>
        <div style="color:var(--text-muted);font-size:13px">Rolled Back</div>
      </div>
      <div class="card" style="text-align:center">
        <div style="font-size:28px;font-weight:700;color:var(--danger,#ef4444)">${stats.failed || 0}</div>
        <div style="color:var(--text-muted);font-size:13px">Failed</div>
      </div>
    </div>`;

  const content = `
    <div class="page-header">
      <h1>Journal</h1>
      <p>Track and rollback agent actions with a complete audit trail</p>
    </div>
    ${statsHtml}
    <div class="card">
      <h3>Journal Entries (${entries.length})</h3>
      ${table}
    </div>`;

  res.send(layout('journal', req.session.user, content, flash));
});

router.post('/journal/:id/rollback', requireAuth, async (req, res) => {
  const result = await apiPost(`/engine/journal/${req.params.id}/rollback`, req.session.token, {});

  if (result.status < 300) {
    req.session.flash = { message: 'Action rolled back successfully', type: 'success' };
  } else {
    req.session.flash = { message: (result.body && result.body.error) || 'Failed to rollback action', type: 'danger' };
  }
  res.redirect('/journal');
});

module.exports = router;

File: dashboards/go/handlers/journal.go
package handlers

import (
	"agenticmail-dashboard/services"
	"agenticmail-dashboard/templates"
	"fmt"
	"net/http"
)

// HandleJournal handles the journal page (GET) with entries and stats,
// and POST for rollback actions.
func HandleJournal(w http.ResponseWriter, r *http.Request) {
	s := services.GetSession(r)

	if r.Method == "POST" {
		r.ParseForm()
		action := r.FormValue("action")
		if action == "rollback" {
			services.APICall("/engine/journal/"+r.FormValue("id")+"/rollback", "POST", s.Token, nil)
		}
		http.Redirect(w, r, "/journal", http.StatusFound)
		return
	}

	entries, _ := services.APICall("/engine/journal", "GET", s.Token, nil)
	stats, _ := services.APICall("/engine/journal/stats/default", "GET", s.Token, nil)
	if stats == nil {
		stats = map[string]interface{}{}
	}

	var statsHTML string
	statsHTML = fmt.Sprintf(`<div class="stats">
<div class="stat"><div class="l">Total Entries</div><div class="v" style="color:var(--primary)">%d</div></div>
<div class="stat"><div class="l">Actions Logged</div><div class="v" style="color:var(--success)">%d</div></div>
<div class="stat"><div class="l">Rollbacks</div><div class="v" style="color:var(--warning)">%d</div></div></div>`,
		templates.IntVal(stats, "totalEntries"), templates.IntVal(stats, "totalActions"), templates.IntVal(stats, "totalRollbacks"))

	var tableHTML string
	if entryList, ok := entries["entries"].([]interface{}); ok && len(entryList) > 0 {
		rows := ""
		for _, en := range entryList {
			e := en.(map[string]interface{})
			rollbackBtn := ""
			if templates.StrVal(e, "status") != "rolled_back" {
				rollbackBtn = fmt.Sprintf(`<form method="POST" action="/journal" style="display:inline"><input type="hidden" name="action" value="rollback"><input type="hidden" name="id" value="%s"><button class="btn btn-sm btn-d" type="submit">Rollback</button></form>`, templates.Esc(e["id"]))
			}
			rows += fmt.Sprintf(`<tr><td style="font-weight:600">%s</td><td style="color:var(--dim)">%s</td><td>%s</td><td style="font-size:12px;color:var(--muted)">%s</td><td>%s</td></tr>`,
				templates.Esc(e["action"]), templates.Esc(e["agent"]), templates.Badge(templates.StrVal(e, "status")), templates.Esc(e["timestamp"]), rollbackBtn)
		}
		tableHTML = `<table><thead><tr><th>Action</th><th>Agent</th><th>Status</th><th>Time</th><th></th></tr></thead><tbody>` + rows + `</tbody></table>`
	} else {
		tableHTML = `<div class="empty"><div class="empty-i">📓</div>No journal entries yet</div>`
	}

	content := fmt.Sprintf(`<h2 class="t">Journal</h2><p class="desc">Immutable action log with rollback capability</p>
%s<div class="card"><div class="ct">Journal Entries</div>%s</div>`, statsHTML, tableHTML)

	w.Header().Set("Content-Type", "text/html")
	fmt.Fprint(w, templates.Layout("journal", s.User, content))
}
